File: android/swapl/app/src/main/java/app/swapl/core/repository/ProfileRepository.kt
package app.swapl.core.repository

import app.swapl.core.model.InterestsCatalog
import app.swapl.core.model.MeResponse
import app.swapl.core.model.PublicProfile
import app.swapl.core.model.SavedSearch
import app.swapl.core.model.SavedSearchesResponse
import app.swapl.core.model.UserSettings
import app.swapl.core.network.ApiClient
import io.ktor.client.call.body
import io.ktor.client.request.get
import io.ktor.client.request.patch
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.http.ContentType
import io.ktor.http.contentType
import kotlinx.serialization.Serializable
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class ProfileRepository @Inject constructor(private val api: ApiClient) {

    suspend fun me(): MeResponse =
        api.client.get("${api.baseUrl}/api/me").body()

    suspend fun publicProfile(id: String): PublicProfile =
        api.client.get("${api.baseUrl}/api/profiles/$id").body()

    // PATCH /api/profile — partial update (DOK-147). `name` is skipped when
    // null (the API requires a non-empty name); the other strings are always
    // sent so empty input clears the nullable fields server-side.
    suspend fun updateProfile(body: ProfileUpdateBody) {
        api.client.patch("${api.baseUrl}/api/profile") {
            contentType(ContentType.Application.Json)
            setBody(body)
        }
    }

    // GET/PATCH /api/profile/settings — privacy & notification toggles.
    suspend fun settings(): UserSettings =
        api.client.get("${api.baseUrl}/api/profile/settings")
            .body<SettingsResponse>().settings

    // Partial merge server-side: omitted (null) keys keep their stored value.
    suspend fun updateSettings(patch: SettingsPatch): UserSettings =
        api.client.patch("${api.baseUrl}/api/profile/settings") {
            contentType(ContentType.Application.Json)
            setBody(patch)
        }.body<SettingsResponse>().settings

    suspend fun interests(): InterestsCatalog =
        api.client.get("${api.baseUrl}/api/profile/interests").body()

    suspend fun saveInterests(slugs: List<String>, bioVibe: String?) =
        api.client.post("${api.baseUrl}/api/profile/interests") {
            contentType(ContentType.Application.Json)
            setBody(InterestsBody(slugs, bioVibe))
        }

    // POST /api/auth/change-password (DOK-149). currentPassword is null when a
    // social/OTP-only account sets its first password; the server keeps THIS
    // device's token valid and revokes every other one.
    suspend fun changePassword(currentPassword: String?, newPassword: String) {
        api.client.post("${api.baseUrl}/api/auth/change-password") {
            contentType(ContentType.Application.Json)
            setBody(ChangePasswordBody(currentPassword, newPassword))
        }
    }

    suspend fun savedSearches(): List<SavedSearch> =
        api.client.get("${api.baseUrl}/api/saved-searches").body<SavedSearchesResponse>().items

    suspend fun report(reason: String, detail: String?, listingId: String?, targetUserId: String?) =
        api.client.post("${api.baseUrl}/api/reports") {
            contentType(ContentType.Application.Json)
            setBody(ReportBody(reason, detail, listingId, targetUserId))
        }

    @Serializable
    data class ProfileUpdateBody(
        // Defaulted to null so kotlinx-serialization omits it from the JSON
        // (encodeDefaults=false) — the stored name is kept rather than rejected.
        val name: String? = null,
        val bio: String,
        val work: String,
        val languages: List<String>,
        val homeCity: String,
        val homeCountry: String,
    )

    @Serializable
    data class SettingsPatch(
        val searchEngineIndexing: Boolean? = null,
        val showHomeCity: Boolean? = null,
        val emailNotifications: Boolean? = null,
        val pushNotifications: Boolean? = null,
    )

    @Serializable
    private data class ChangePasswordBody(
        val currentPassword: String? = null,
        val newPassword: String,
    )

    @Serializable
    private data class SettingsResponse(val settings: UserSettings)

    @Serializable
    private data class InterestsBody(
        val interests: List<String>,
        val bioVibe: String? = null,
    )

    @Serializable
    private data class ReportBody(
        val reason: String,
        val detail: String? = null,
        val listingId: String? = null,
        val targetUserId: String? = null,
    )
}

File: ios/Swapl/Core/Repositories/ProfileRepository.swift
import Foundation

final class ProfileRepository: @unchecked Sendable {
    static let shared = ProfileRepository()

    func publicProfile(id: String) async throws -> PublicProfile {
        try await APIClient.shared.send("GET", "/api/profiles/\(id)")
    }

    // Current-user snapshot — used by the Personal information editor to
    // prefill fields and by the settings screens for the initial toggles.
    func me() async throws -> MeResponse {
        try await APIClient.shared.send("GET", "/api/me")
    }

    // PATCH /api/profile — partial update; only the non-nil keys are sent.
    // Strings are nullable server-side so empty input clears the field.
    struct ProfileUpdateBody: Encodable {
        var name: String?
        var bio: String?
        var work: String?
        var languages: [String]?
        var homeCity: String?
        var homeCountry: String?
        // Off-platform contact channels (DOK-204). Full-replace: the complete
        // desired set is sent each save; the server normalizes + drops invalids.
        var contactChannels: ContactChannels?
    }

    func updateProfile(_ body: ProfileUpdateBody) async throws -> EmptyResponse {
        try await APIClient.shared.send("PATCH", "/api/profile", body: body)
    }

    // GET/PATCH /api/profile/settings — privacy & notification toggles.
    private struct SettingsResponse: Decodable { let settings: UserSettings }

    func settings() async throws -> UserSettings {
        let r: SettingsResponse = try await APIClient.shared.send("GET", "/api/profile/settings")
        return r.settings
    }

    // Partial merge server-side: omitted keys keep their stored value.
    struct SettingsPatch: Encodable {
        var searchEngineIndexing: Bool?
        var showHomeCity: Bool?
        var emailNotifications: Bool?
        var pushNotifications: Bool?
        var countDaysAbroad: Bool?
    }

    func updateSettings(_ patch: SettingsPatch) async throws -> UserSettings {
        let r: SettingsResponse = try await APIClient.shared.send("PATCH", "/api/profile/settings", body: patch)
        return r.settings
    }

    func interests() async throws -> InterestsCatalog {
        try await APIClient.shared.send("GET", "/api/profile/interests")
    }

    struct UpdateInterestsBody: Encodable {
        let interests: [String]
        let bioVibe: String?
    }

    func updateInterests(_ body: UpdateInterestsBody) async throws -> EmptyResponse {
        try await APIClient.shared.send("POST", "/api/profile/interests", body: body)
    }

    func savedSearches() async throws -> [SavedSearch] {
        struct Response: Decodable { let items: [SavedSearch] }
        let r: Response = try await APIClient.shared.send("GET", "/api/saved-searches")
        return r.items
    }

    struct ReportBody: Encodable {
        let reason: String
        let detail: String?
        let listingId: String?
        let targetUserId: String?
    }

    func report(_ body: ReportBody) async throws -> EmptyResponse {
        try await APIClient.shared.send("POST", "/api/reports", body: body)
    }
}
